js/ui/grid_core/ui.grid_core.utils.js

```diff
@@ -544,5 +547,53 @@ export default {
         }
 
         return 15000000 / this.getPixelRatio(getWindow());
-    }
+    },
+
+    normalizeLookupDataSource(lookup) {
+        let lookupDataSourceOptions;
+        if(lookup.items) {
+            lookupDataSourceOptions = lookup.items;
+        } else {
+            lookupDataSourceOptions = lookup.dataSource;
+            if(isFunction(lookupDataSourceOptions) && !variableWrapper.isWrapped(lookupDataSourceOptions)) {
+                lookupDataSourceOptions = lookupDataSourceOptions({});
+            }
+        }
+
+        return normalizeDataSourceOptions(lookupDataSourceOptions);
+    },
+
+    getWrappedLookupDataSource(column, dataSource, filter) {
+        const lookupDataSourceOptions = this.normalizeLookupDataSource(column.lookup);
+
+        const lookupDataSource = {
+            load: (loadOptions) => {
+                const d = new Deferred();
+                dataSource.load({
+                    ...loadOptions,
+                    filter: this.combineFilters([filter, loadOptions.filter], 'and'),
+                    group: column.dataField || column.name
+                }).done((items) => {
+                    const filter = this.combineFilters(
+                        items.map((data => [
+                            column.lookup.valueExpr, data.key,
+                        ])),
+                        'or'
+                    );
+
+                    (new DataSource({
+                        ...lookupDataSourceOptions,
```

Comment:
It is possible to use column.lookup.items here if it is defined to avoid additional request:
`...(column.lookup.items ? {.store: column.lookup.items } : lookupDataSourceOptions)`
If lookup optimisation is not applied (https://js.devexpress.com/Documentation/Guide/UI_Components/DataGrid/Enhance_Performance_on_Large_Datasets/#Lookup_Optimization) DataGrid loads lookup on first render and loaded lookup items are assigned to lookup.items.

Reply:
If filter is big server can return get request length limit error. For lookup column with optimisation when calculateDisplayValue field is defined we can make filtering by this display column to avoid additional request with big filter. We should discuss this. Or showRelevantValues should be disabled by default. It will be bad if after upgrading some customers will see request length limit error.